plugins/modules/ipaautomountkey.py

```diff
@@ -0,0 +1,217 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+ANSIBLE_METADATA = {
+    "metadata_version": "1.0",
+    "supported_by": "community",
+    "status": ["preview"],
+}
+
+
+DOCUMENTATION = '''
+---
+module: ipaautomountkey
+author: chris procter
+short_description: Manage FreeIPA autommount map
+description:
+- Add, delete, and modify an IPA automount map
+options:
+  ipaadmin_principal:
+    description: The admin principal
+    default: admin
+  ipaadmin_password:
+    description: The admin password
+    required: False
+  location:
+    description: automount location map is in
+    required: True
+    choices: ["automountlocationcn", "automountlocation"]
+  mapname:
+    description: automount map to be managed
+    choices: ["map", "automountmapname", "automountmap"]
+    required: True
+  key:
+    description: automount key to be managed
+    required: True
+    choices: ["name", "automountkey"]
+  newkey:
+    description: key to change to if state=rename
+    required: True
+    choices: ["newname", "newautomountkey"]
+  info:
+    description: Mount information for the key
+    required: True
+    choices: ["information", "newinfo", "automountinformation"]
+  state:
+    description: State to ensure
+    required: False
+    default: present
+    choices: ["present", "absent", "rename"]
+'''
+
+EXAMPLES = '''
+  - name: create key TestKey
+    ipaautomountkey:
+      ipaadmin_password: SomeADMINpassword
+      locationcn: TestLocation
+      mapname: TestMap
+      key: TestKey
+      info: 192.168.122.1:/exports
+      state: present
+
+  - name: ensure key TestKey is absent
+    ipaautomountkey:
+      ipaadmin_password: SomeADMINpassword
+      location: TestLocation
+      mapname: TestMap
+      key: TestKey
+      state: absent
+'''
+
+RETURN = '''
+'''
+
+from ansible.module_utils.ansible_freeipa_module import (
+    FreeIPABaseModule, ipalib_errors
+)
+
+
+class AutomountKey(FreeIPABaseModule):
+
+    ipa_param_mapping = {
+        'automountkey': "key",
+        'automountmapautomountmapname': "mapname",
+    }
+
+    def get_key(self, location, mapname, keyname):
+        resp = dict()
+        try:
+            resp = self.api_command("automountkey_show",
+                                    location,
+                                    {"automountmapautomountmapname": mapname,
+                                     "automountkey": keyname})
+        except ipalib_errors.NotFound:
+            pass
+
+        return resp.get("result", None)
+
+    def check_ipa_params(self):
```

Comment:
`info` and `new_name` should are invalid when `state` is `absent`. `new_name` is only valid when `state` is `renamed`.

These should be reflected on this function. (Check [dnszone check_ipa_params()](https://github.com/freeipa/ansible-freeipa/blob/0a1f289f3cd2e4186cae8d7ac933605b16c97b0c/plugins/modules/ipadnszone.py#L441) for the proper fail message.).

Reply:
I disagree, info and new_name are ignored when state=absent, and I'm fine with that. Deliberately locking out keys that are just going to be ignored makes life harder for the user, in this case it means switching a task from state=present to state=absent also requires you to figure out which other keys also need dropping and remove those as well, and then you need to replace them when you switch back to state=present.  This may not seem a very common user case but if you typo your key name then just being able to make a small tweak and rerun the playbook is useful.   So this is extra code that only exists to remove functionality from the user.

Reply:
Other ansible-freeipa modules do not allow the use of variables with `state: absent` (unless on `action: member`). So, raising an error with invalid variables for the selected state will make it consistent with other modules.

In your provided example, a user would still have to ensure the key with typo is absent, and then ensure that the correct key is present. It would still requires two tasks, and one of then don't need the variables (`state: absent`).

I don't see any advantage in allowing the use of variables for `state: absent`.